File: java/Dson-Codec/src/main/java/cn/wjybxx/dsoncodec/DefaultDsonConverter.java
/*
 * Copyright 2023-2024 wjybxx(845740757@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.wjybxx.dsoncodec;

import cn.wjybxx.base.TypeInfo;
import cn.wjybxx.base.io.StringBuilderWriter;
import cn.wjybxx.dson.*;
import cn.wjybxx.dson.io.*;
import cn.wjybxx.dson.text.DsonTextReader;
import cn.wjybxx.dson.text.DsonTextWriter;
import cn.wjybxx.dson.text.ObjectStyle;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.io.Reader;
import java.io.Writer;
import java.util.Arrays;
import java.util.Objects;
import java.util.function.Supplier;

/**
 * @author wjybxx
 * date 2023/4/2
 */
class DefaultDsonConverter implements DsonConverter {

    private final TypeMetaRegistry typeMetaRegistry;
    private final DsonCodecRegistry codecRegistry;
    private final GenericHelper genericHelper;
    private final TypeWriteHelper typeWriteHelper;
    private final ConverterOptions options;

    DefaultDsonConverter(TypeMetaRegistry typeMetaRegistry,
                         DsonCodecRegistry codecRegistry,
                         GenericHelper genericHelper,
                         TypeWriteHelper typeWriteHelper,
                         ConverterOptions options) {
        this.codecRegistry = codecRegistry;
        this.typeMetaRegistry = typeMetaRegistry;
        this.genericHelper = genericHelper;
        this.typeWriteHelper = typeWriteHelper;
        this.options = options;
    }

    @Override
    public DsonCodecRegistry codecRegistry() {
        return codecRegistry;
    }

    @Override
    public TypeMetaRegistry typeMetaRegistry() {
        return typeMetaRegistry;
    }

    @Override
    public GenericHelper genericCodecHelper() {
        return genericHelper;
    }

    @Override
    public ConverterOptions options() {
        return options;
    }

    @Override
    public DsonConverter withOptions(ConverterOptions options) {
        Objects.requireNonNull(options);
        return new DefaultDsonConverter(typeMetaRegistry, codecRegistry, genericHelper, typeWriteHelper, options);
    }
    // region binary

    @Nonnull
    @Override
    public byte[] write(Object value, @Nonnull TypeInfo declaredType) {
        Objects.requireNonNull(value);
        // 外部销毁，确保buffer归还
        try (var outputStream = DsonOutputs.newInstance(options.bufferPool, options.bufferLength, options.maxBufferLength)) {
            encodeObject(outputStream, value, declaredType);
            return Arrays.copyOfRange(outputStream.getBuffer(), 0, outputStream.getPosition());
        }
    }

    @Override
    public <T> T read(byte[] source, @Nonnull TypeInfo declaredType, Supplier<? extends T> factory) {
        try (DsonInput inputStream = DsonInputs.newInstance(source)) {
            return decodeObject(inputStream, declaredType, factory);
        }
    }

    @Override
    public void write(Object value, TypeInfo declaredType, DsonChunk chunk) {
        Objects.requireNonNull(value);
        try (DsonOutput outputStream = DsonOutputs.newInstance(chunk.getBuffer(), chunk.getOffset(), chunk.getLength())) {
            encodeObject(outputStream, value, declaredType);
            chunk.setUsed(outputStream.getPosition());
        }
    }

    @Override
    public <T> T read(DsonChunk chunk, TypeInfo declaredType, Supplier<? extends T> factory) {
        try (DsonInput inputStream = DsonInputs.newInstance(chunk.getBuffer(), chunk.getOffset(), chunk.getLength())) {
            T result = decodeObject(inputStream, declaredType, factory);
            chunk.setUsed(inputStream.getPosition());
            return result;
        }
    }

    @Override
    public void write(Object value, TypeInfo declaredType, DsonOutput output) {
        Objects.requireNonNull(value);
        Objects.requireNonNull(output, "output");
        encodeObject(output, value, declaredType);
    }

    @Override
    public <T> T read(DsonInput input, TypeInfo declaredType, @Nullable Supplier<? extends T> factory) {
        Objects.requireNonNull(input, "input");
        return decodeObject(input, declaredType, factory);
    }

    @Override
    public <T> T cloneObject(Object value, TypeInfo declaredType, TypeInfo targetType, Supplier<? extends T> factory) {
        if (value == null) return null;
        try (var outputStream = DsonOutputs.newInstance(options.bufferPool, options.bufferLength, options.maxBufferLength)) {
            encodeObject(outputStream, value, declaredType);
            // 不销毁
            DsonInput inputStream = DsonInputs.newInstance(outputStream.getBuffer(), 0, outputStream.getPosition());
            return decodeObject(inputStream, targetType, factory);
        }
    }

    /** 注意：由外部销毁输出流 */
    private void encodeObject(DsonOutput outputStream, Object value, TypeInfo typeInfo) {
        DsonBinaryWriter binaryWriter = new DsonBinaryWriter(options.binWriterSettings, outputStream, false);
        try (DsonObjectWriter wrapper = new DefaultDsonObjectWriter(this, typeWriteHelper, binaryWriter)) {
            wrapper.writeObject(null, value, typeInfo, null);
            wrapper.flush();
        }
    }

    /** 注意：由外部销毁输入流 */
    private <T> T decodeObject(DsonInput inputStream, TypeInfo typeInfo, Supplier<? extends T> factory) {
        DsonBinaryReader binaryReader = new DsonBinaryReader(options.binReaderSettings, inputStream, false);
        try (DsonObjectReader wrapper = wrapReader(binaryReader)) {
            return wrapper.readObject(null, typeInfo, factory);
        }
    }

    private DsonObjectReader wrapReader(DsonReader reader) {
        return new DefaultDsonObjectReader(this, toDsonCollectionReader(reader));
    }

    private DsonCollectionReader toDsonCollectionReader(DsonReader dsonReader) {
        assert !(dsonReader instanceof DsonCollectionReader);
        DsonValue dsonValue = Dsons.readTopDsonValue(dsonReader);
        return DsonCollectionReader.unsafeCreate(options.binReaderSettings, dsonValue, true);
    }
    // endregion

    // region text
    @Nonnull
    @Override
    public String writeAsDson(Object value, @Nonnull TypeInfo declaredType, ObjectStyle style) {
        StringBuilder stringBuilder = options.stringBuilderPool.acquire();
        try {
            writeAsDson(value, declaredType, new StringBuilderWriter(stringBuilder), style);
            return stringBuilder.toString();
        } finally {
            options.stringBuilderPool.release(stringBuilder);
        }
    }

    @Override
    public <T> T readFromDson(CharSequence source, @Nonnull TypeInfo declaredType, Supplier<? extends T> factory) {
        try (DsonObjectReader wrapper = wrapReader(new DsonTextReader(options.textReaderSettings, source))) {
            return wrapper.readObject(null, declaredType, factory);
        }
    }

    @Override
    public void writeAsDson(Object value, @Nonnull TypeInfo declaredType, Writer writer, ObjectStyle style) {
        Objects.requireNonNull(writer, "writer");
        try (DsonObjectWriter wrapper = new DefaultDsonObjectWriter(this, typeWriteHelper,
                new DsonTextWriter(options.textWriterSettings, writer, false))) {
            wrapper.writeObject(null, value, declaredType, style);
            wrapper.flush();
        }
    }

    @Override
    public <T> T readFromDson(Reader source, @Nonnull TypeInfo declaredType, Supplier<? extends T> factory) {
        try (DsonObjectReader wrapper = wrapReader(
                new DsonTextReader(options.textReaderSettings, Dsons.newStreamScanner(source, false)))) {
            return wrapper.readObject(null, declaredType, factory);
        }
    }

    @Override
    public DsonValue writeAsDsonValue(Object value, TypeInfo declaredType) {
        Objects.requireNonNull(value);
        DsonArray<String> outList = new DsonArray<>(1);
        try (DsonObjectWriter wrapper = new DefaultDsonObjectWriter(this, typeWriteHelper,
                new DsonCollectionWriter(options.binWriterSettings, outList))) {
            wrapper.writeObject(null, value, declaredType, ObjectStyle.INDENT);
            DsonValue dsonValue = outList.get(0);
            if (dsonValue.getDsonType().isContainer()) {
                return dsonValue;
            }
            throw new IllegalArgumentException("value must be container");
        }
    }

    @Override
    public <T> T readFromDsonValue(DsonValue source, @Nonnull TypeInfo declaredType, Supplier<? extends T> factory) {
        if (!source.getDsonType().isContainer()) {
            throw new IllegalArgumentException("value must be container");
        }
        try (DsonObjectReader wrapper = new DefaultDsonObjectReader(this,
                DsonCollectionReader.unsafeCreate(options.binReaderSettings, source, true))) {
            return wrapper.readObject(null, declaredType, factory);
        }
    }

    @Override
    public DsonValue readAsDsonValue(Reader source) {
        try (DsonReader textReader = new DsonTextReader(options.textReaderSettings, Dsons.newStreamScanner(source, false))) {
            return Dsons.readTopDsonValue(textReader);
        }
    }

    @Override
    public DsonValue readAsDsonValue(DsonInput source) {
        try (DsonReader binaryReader = new DsonBinaryReader(options.binReaderSettings, source, false)) {
            return Dsons.readTopDsonValue(binaryReader);
        }
    }

    // endregion
}

File: csharp/Wjybxx.Dson.Codec/src/DefaultDsonConverter.cs
#region LICENSE

// Copyright 2024 wjybxx(845740757@qq.com)
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#endregion

using System;
using System.Diagnostics;
using System.IO;
using System.Runtime.CompilerServices;
using System.Text;
using Wjybxx.Commons;
using Wjybxx.Commons.Attributes;
using Wjybxx.Dson.IO;
using Wjybxx.Dson.Text;

namespace Wjybxx.Dson.Codec
{
/// <summary>
/// 默认实现
/// </summary>
[ThreadSafe]
public class DefaultDsonConverter : IDsonConverter
{
    private readonly DynamicTypeMetaRegistry typeMetaRegistry;
    private readonly DynamicCodecRegistry codecRegistry;
    private readonly TypeWriteHelper typeWriteHelper;
    private readonly ConverterOptions options;

    internal DefaultDsonConverter(DynamicTypeMetaRegistry typeMetaRegistry,
                                  DynamicCodecRegistry codecRegistry,
                                  TypeWriteHelper typeWriteHelper,
                                  ConverterOptions options) {
        this.typeMetaRegistry = typeMetaRegistry;
        this.codecRegistry = codecRegistry;
        this.typeWriteHelper = typeWriteHelper;
        this.options = options;
    }

    #region other

    public ConverterOptions Options => options;
    ITypeMetaRegistry IDsonConverter.TypeMetaRegistry => typeMetaRegistry;
    IDsonCodecRegistry IDsonConverter.CodecRegistry => codecRegistry;

    public IDsonConverter WithOptions(ConverterOptions options) {
        if (options == null) throw new ArgumentNullException(nameof(options));
        return new DefaultDsonConverter(typeMetaRegistry, codecRegistry, typeWriteHelper, options);
    }

    /// <summary>
    /// 暴露注册表以允许用户提前缓存
    /// </summary>
    public DynamicTypeMetaRegistry TypeMetaRegistry => typeMetaRegistry;

    /// <summary>
    /// 暴露注册表以允许用户提前缓存
    /// </summary>
    public DynamicCodecRegistry CodecRegistry => codecRegistry;

    #endregion

    #region binary

    public byte[] Write(object value, Type declaredType) {
        if (value == null) throw new ArgumentNullException(nameof(value));
        if (declaredType == null) throw new ArgumentNullException(nameof(declaredType));
        // 外部销毁流，确保buffer规划到池
        using var outputStream = DsonOutputs.NewInstance(options.bufferPool, options.bufferLength, options.maxBufferLength);
        EncodeObject(outputStream, value, declaredType);
        return ArrayUtil.CopyOf(outputStream.Buffer, 0, outputStream.Position);
    }

    public object Read(byte[] source, Type declaredType, Func<object>? factory = null) {
        using IDsonInput inputStream = DsonInputs.NewInstance(source);
        return DecodeObject(inputStream, declaredType, factory);
    }

    public void Write(object value, Type declaredType, DsonChunk chunk) {
        if (value == null) throw new ArgumentNullException(nameof(value));
        using IDsonOutput outputStream = DsonOutputs.NewInstance(chunk.Buffer, chunk.Offset, chunk.Length);
        EncodeObject(outputStream, value, declaredType);
        chunk.Used = outputStream.Position;
    }

    public object Read(DsonChunk chunk, Type declaredType, Func<object>? factory = null) {
        using IDsonInput inputStream = DsonInputs.NewInstance(chunk.Buffer, chunk.Offset, chunk.Length);
        object result = DecodeObject(inputStream, declaredType, factory);
        chunk.Used = inputStream.Position;
        return result;
    }

    public object CloneObject(object? value, Type declaredType, Type targetType, Func<object>? factory = null) {
        if (value == null) return null!;
        if (value.GetType().IsValueType) return value;

        using var dsonOutput = DsonOutputs.NewInstance(options.bufferPool, options.bufferLength, options.maxBufferLength);
        EncodeObject(dsonOutput, value, declaredType);
        // 不销毁
        IDsonInput inputStream = DsonInputs.NewInstance(dsonOutput.Buffer, 0, dsonOutput.Position);
        return DecodeObject(inputStream, targetType, factory);
    }

    /** 注意：由外部销毁输出流 */
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private void EncodeObject(IDsonOutput outputStream, object value, Type declaredType) {
        DsonBinaryWriter<string> binaryWriter = new DsonBinaryWriter<string>(options.binWriterSettings, outputStream, autoClose: false);
        using DefaultDsonObjectWriter wrapper = new DefaultDsonObjectWriter(this, typeWriteHelper, binaryWriter);
        wrapper.WriteObject(null, value, declaredType);
        wrapper.Flush();
    }

    /** 注意：由外部销毁输入流 */
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private object DecodeObject(IDsonInput inputStream, Type declaredType, Func<object>? factory) {
        DsonBinaryReader<string> binaryReader = new DsonBinaryReader<string>(options.binReaderSettings, inputStream, autoClose: false);
        using IDsonObjectReader wrapper = WrapReader(binaryReader);
        return wrapper.ReadObject(null, declaredType, factory);
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private IDsonObjectReader WrapReader(IDsonReader<string> reader) {
        if (options.randomRead) {
            return new BufferedDsonObjectReader(this, ToDsonCollectionReader(reader));
        } else {
            return new DefaultDsonObjectReader(this, reader);
        }
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private DsonCollectionReader<string> ToDsonCollectionReader(IDsonReader<string> dsonReader) {
        Debug.Assert(dsonReader is not DsonCollectionReader<string>);
        // 如果要优化gc的话，需要传入DsonObject和DsonArray的对象池... 这和外部缓存DsonValue是两个优化
        DsonValue dsonValue = Dsons.ReadTopDsonValue(dsonReader) ?? throw new DsonCodecException("eof");
        return DsonCollectionReader<string>.UnsafeCreate(options.binReaderSettings, dsonValue);
    }

    #endregion

    #region text

    public string WriteAsDson(object value, Type declaredType, ObjectStyle? style = null) {
        if (value == null) throw new ArgumentNullException(nameof(value));
        StringBuilder stringBuilder = options.stringBuilderPool.Acquire();
        try {
            WriteAsDson(value, declaredType, new StringWriter(stringBuilder), style);
            return stringBuilder.ToString();
        }
        finally {
            options.stringBuilderPool.Release(stringBuilder);
        }
    }

    public object ReadFromDson(string source, Type declaredType, Func<object>? factory = null) {
        DsonTextReader textReader = new DsonTextReader(options.textReaderSettings, source);
        using IDsonObjectReader wrapper = WrapReader(textReader);
        return wrapper.ReadObject(null, declaredType, factory);
    }

    public void WriteAsDson(object value, Type declaredType, TextWriter writer, ObjectStyle? style = null) {
        if (value == null) throw new ArgumentNullException(nameof(value));
        if (writer == null) throw new ArgumentNullException(nameof(writer));

        DsonTextWriter textWriter = new DsonTextWriter(options.textWriterSettings, writer, false);
        using DefaultDsonObjectWriter wrapper = new DefaultDsonObjectWriter(this, typeWriteHelper, textWriter);
        wrapper.WriteObject(null, value, declaredType, style);
        wrapper.Flush();
    }

    public object ReadFromDson(TextReader source, Type declaredType, Func<object>? factory = null) {
        DsonTextReader textReader = new DsonTextReader(options.textReaderSettings, Dsons.NewStreamScanner(source, false));
        using IDsonObjectReader wrapper = WrapReader(textReader);
        return wrapper.ReadObject(null, declaredType, factory);
    }

    public DsonValue WriteAsDsonValue(object value, Type declaredType) {
        if (value == null) throw new ArgumentNullException(nameof(value));
        DsonArray<string> outList = new DsonArray<string>(1);
        IDsonWriter<string> objectWriter = new DsonCollectionWriter<string>(options.binWriterSettings, outList);
        using IDsonObjectWriter wrapper = new DefaultDsonObjectWriter(this, typeWriteHelper, objectWriter);

        wrapper.WriteObject(null, value, declaredType, ObjectStyle.Flow);
        DsonValue dsonValue = outList[0];
        if (dsonValue.DsonType.IsContainer()) {
            return dsonValue;
        }
        throw new AggregateException("value must be container");
    }

    public object ReadFromDsonValue(DsonValue source, Type declaredType, Func<object>? factory = null) {
        if (!source.DsonType.IsContainer()) {
            throw new ArgumentException("value must be container");
        }
        DsonCollectionReader<string> objectReader =
            DsonCollectionReader<string>.UnsafeCreate(options.binReaderSettings, source);
        using IDsonObjectReader wrapper = new BufferedDsonObjectReader(this, objectReader);
        return wrapper.ReadObject(null, declaredType, factory);
    }

    public DsonValue ReadAsDsonValue(TextReader source) {
        using DsonTextReader textReader = new DsonTextReader(options.textReaderSettings, Dsons.NewStreamScanner(source, false));
        return Dsons.ReadTopDsonValue(textReader)!;
    }

    #endregion
}
}